CaseStudyProductListing/Shared/NetworkManager/NetworkManager.swift

```diff
@@ -0,0 +1,63 @@
+//
+//  NetworkManager.swift
+//  CaseStudyProductListing
+//
+//  Created by Atul Vishnoi on 13/07/23.
+//
+
+import Foundation
+
+enum NetworkError: Error {
+    case invalidUrl
+    case invalidData
+}
+
+enum FeedType: String {
+    case emailed
+    case shared
+    case viewed
+}
+
+struct NetworkManager {
+    // Base API URL
+    private let baseUrl = "https://api.nytimes.com/svc/mostpopular/v2/"
+    // API Key
+    private let apiKey = "RAkAhPXo0kaO0yqbS8W6ZfshjhjvZft8"
+
+    typealias FeedCompletionClosure = ((ArticlesResponse?, Error?) -> Void)
+
+    private func createRequest(for url: String) -> URLRequest? {
+        guard let url = URL(string: url) else { return nil }
+        var request = URLRequest(url: url)
+        request.httpMethod = "GET"
+        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
+        return request
+    }
+
+    private func executeRequest<T: Codable>(request: URLRequest, completion: ((T?, Error?) -> Void)?) {
+        let session = URLSession(configuration: .default)
+        let dataTask = session.dataTask(with: request) { (data, _, error) in
```

Comment:
Consider using a background queue for the network request and dispatching the completion closure to the appropriate queue depending on the desired behavior and usage of the code.